File: LeetCode/Algo/Day6/permutation_string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()){
            return false;
        }
        int[] array1 = new int[26];
        for(int i = 0; i < s1.length();i++){
            array1[s1.charAt(i) - 'a']++;
        }
        for(int i = 0; i < s2.length()-s1.length()+1;i++){
            int[] array2 = new int[26];
            for(int j = 0; j < s1.length();j++){
                array2[s2.charAt(i+j) - 'a']++;
            }
            if(matches(array1, array2))
                return true;
        }
        return false;
        
    }
    public boolean matches(int[] a1, int[] a2){
        for(int i = 0; i < 26; i++){
            if(a1[i] != a2[i])
                return false;
        }
        return true;
    }
}

File: LeetCode/Algo/Day6/permutation_string.py

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # Sử dụng kĩ thuật mảng đánh dấu 
        if len(s1) > len(s2):
            return False
        array1 = [0] * 26
        for i in s1:
            array1[ord(i) - 97] += 1
        
        for i in range(len(s2)-len(s1)+1):
            array2 = [0] * 26 
            for j in range(len(s1)):
                array2[ord(s2[i+j])-97] += 1
            if self.match(array1, array2):
                return True
        return False
    
    def match(self, array1, array2):
        for i in range(26):
            if array1[i] != array2[i]:
                return False
        return True
        
           
solve = Solution()
print(solve.checkInclusion('ab', 'eidbaooo'))

# https://leetcode.com/problems/permutation-in-string/